### DISCUSSION TITLE
Rendering/concealing tables (and drawers?)

### DISCUSSION THREAD

**Discussion by mrossinek** [nvim-neorg/neorg#17] [Q&A] [answered]
I recently found this project and am already in love with it! Congrats to the devs! :slightly_smiling_face: 

I read the specs and wiki and I have a question about rendering in general (but mostly about tables).
The specs suggest, what a `@table` should look like once it is rendered. First off: how is this to be understood? Is this under the assumption that a Neorg file gets rendered as markdown and/or html and should then produce the same result as the proposed markdown table in the spec?
If yes, does such a renderer already exist?

Furthermore, since you also say that tables are not as powerful as org-mode tables, **yet**, what are your plans/ideas for tables in the future? Will they get some syntactic sugar via e.g. some detached modifier syntax? Or will they always be abstracted away via the `@table` tag? If the latter (which I am fine with) are there any plans to render/conceal a table in-place to make it more syntactically appealing (similar to the rendered Markdown-like example)?
And thinking even further ahead, if concealing a table in-place is a direction you'd like to go towards, are you also thinking of "smart" tables which can do basic arithmetics? This could even be inserted during concealing.. :thinking: 

And finally I have similar questions about the `drawers` (whose concept I was unfamiliar with before). Are those simply `<details></details>` elements from markdown? Then they could simply get folded by default, right? Or is there more to them?

**Comment by vhyrro** [comment, marked as answer]
Great bunch of questions. I'll try answer them one by one:

- Yes, Neorg will have the ability to get rendered as HTML - no, we don't have a renderer :(. The idea is that tables will get rendered just like markdown tables are rendered - it's just that the `@table` tag is a lot cleaner and less difficult to work with for a parser.
- I plan on making tables similar in features to org, hell, we may even go full sc-im style editing at one point :P.
- The concern regarding editing tables is valid, and when designing the specification I've thought of exactly this. Currently conceal isn't powerful enough in Neovim to make the tables *beautified* and things like that. I heard there is a PR coming to Neovim that should make conceals a lot better, but it's not here yet - once it arrives I'll be sure to use it. I do have another idea, however. As soon as insert mode is entered when inside of a `@table` tag you will get transported to a floating window which will be the table editor. This floating window will occupy most of the screen (kind of like a floating terminal). Inside of this window the table will be shown in a way you should be waay more familiar with (using special characters to create table borders etc.). It will literally look like a normal markdown-like or org-like table. Here is where the editing magic will happen. You will have custom keybinds, custom movement and everything you'd need to edit the table; this approach is good in my opinion as it separates editing tables from editing your main document and hence removes the feeling of clutter. After exiting the floating window the `@table` tag will be regenerated and converted from the more complex editor view to the simplified table tag syntax that's present in the spec. Whaddya think?
- Drawers are simply ways to just "hide" away your stuff, so yeah, they're pretty much `<details></details>`. We may add more magic to them at one point, I'm open to suggestions.

**Reply by mrossinek** [reply]
Awesome, thanks for the detailed reply! I was not aware that conceal is not that powerful yet. I just assumed you can write any text on top of any other text :sweat_smile: But I'll be eagerly awaiting that.

Nonetheless, I think the popup for table editing sounds like a good way to go. You're absolutely right that it will make editing a lot cleaner and less cluttered. Especially, as it will simplify handling different mappings when editing table vs normal usage.

And FWIW: I absolutely agree that the `@table` tag is a lot cleaner (especially in terms of parsing) but as you probably will also agree, having a plain-text rendered version of it would help with readability nonetheless.. I'll stay tuned for what you come up with! I don't have too much experience in Lua but in case you ever need help or have something to be looked over, let me know :+1:

**Reply by vhyrro** [reply]
I sure will! As for in-buffer previews we'll probably have to wait for that better conceal unfortunately. I might make a quick keybind at one point that will make the buffer temporarily unwritable and will expand all tables to their ascii versions for quick previews but that's not planned for the close future unfortunately

**Reply by akinsho** [reply]
Despite the limitations of conceal in the short term, you can actually replicate an improved version of conceal by using `extmarks` see `:h nvim_buf_set_extmarks` extmarks can be drawn over characters in a buffer and follow text around. I use them in a few of my plugins [here](https://github.com/akinsho/flutter-tools.nvim#widget-guides-experimental-default-disabled) to draw guides inside a buffer and [here](https://github.com/akinsho/org-bullets.nvim) to replicate conceal/improve its functionality. I guess it's a question of which might be the more complex approach since with extmarks you have to keep track of them and update them as the text changes if necessary.

**Reply by vhyrro** [reply]
Woah, extmarks can override existing text? That's the first time I've heard that. Is that even mentioned anywhere explicitly in the help pages? I remember reading about extmarks, thinking "this might come in handy one day", then leaving it without knowing it could create "true" virtual text. I'll definitely start using those instead of conceal in that case. Since you won't be editing a table "raw" (but through an editor) i think that it shouldn't pose an *insane* amount of headaches, since the underlying text will not be changed live in the buffer.

**Reply by akinsho** [reply]
Yep, it's kind of been a big feature that has been bubbling away in the background since they seem to have big plans for it like adding whole virtual lines to documents and virtual words to lines and all sorts of things so definitely one I watch closely. I don't think it has been announced massively yet as they are just incrementally adding functionality by they I mean (just Bfredl), but I've found it quite useful. Definitely think you (plugin dev) can and they (nvim) might re-implement conceal using it.
